### PR TITLE
Add core tensorflow 2.x api example to mlflow/examples

### LINKED ISSUE #8049: [FR] Add Core TensorFlow Example to mlflow/examples
### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

With MLflow 2.x, we removed the seriously outdated TF1.x example (due to relevance issues). We'd like to solicit a new example from the community of fantastic TF users to create a replacement example that leverages purely TF core APIs (no Keras usage). 

### Motivation

> #### What is the use case for this feature?
To help users with a utilization example that uses the TF core library.

> #### Why is this use case valuable to support for MLflow users in general?
There are MLflow users that leverage base core TF library to build projects. It would be great to have an up to date example on how to leverage the interaction with MLflow APIs.

> #### Why is this use case valuable to support for your project(s) or organization?
This is helpful for TF users who do not use Keras for a project.

> #### Why is it currently difficult to achieve this use case?
We removed the old (arguably irrelevant) 1.x examples with the 2.x release of MLflow.

### Details

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [X] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by dheerajnbhat**
## Related Issues/PRs

Resolve [#8049](https://github.com/mlflow/mlflow/issues/8049#issue-1628494670)  - Add Core TensorFlow Example to mlflow/examples

## What changes are proposed in this pull request?

Added core tensorflow 2.x api example to mlflow/examples

## How is this patch tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

Modified tensorflow version in examples/README.md

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add usage examples for `mlflow.tensorflow` APIs for tensorflow 2.x

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: examples/tensorflow/conda.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+channels:
+  - conda-forge
+  - defaults
+dependencies:
+  - python=3.8
+  - pip
+  - pip:
+      - mlflow>=2.0
+      - tensorflow>=2.8
+      - pandas
```
Comment (harupy) [inline examples/tensorflow/conda.yaml:11]:
```suggestion
      - pandas

```

File: examples/tensorflow/train.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
+# tensorflow 2.x core api
```
Comment (harupy) [inline examples/tensorflow/train.py:1]: How long does this example take to run?

File: examples/tensorflow/train.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        if tracking_url_type_store != "file":
+            # Register the model
+            # There are other ways to use the Model Registry, which depends on the use case,
+            # please refer to the doc for more information:
+            # https://mlflow.org/docs/latest/model-registry.html#api-workflow
+            mlflow.tensorflow.log_model(
+                lin_reg_export, "model", registered_model_name="LinearRegAutoMpgModel"
+            )
+        else:
+            mlflow.tensorflow.log_model(lin_reg_export, "model")
```
Comment (harupy) [inline examples/tensorflow/train.py:200]:
```suggestion
        # I think we can just log the model.
        mlflow.tensorflow.log_model(lin_reg_export, "model")
```

File: examples/tensorflow/train.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        mlflow.log_metric("final_test_loss", test_loss)
+        print(f"\nFinal train loss: {train_loss:0.3f}")
+        print(f"Final test loss: {test_loss:0.3f}")
+
+        # Export the tensorflow model
+        lin_reg_export = ExportModule(
+            model=lin_reg, extract_features=onehot_origin, norm_x=norm_x, norm_y=norm_y
+        )
+
+        tracking_url_type_store = urlparse(mlflow.get_tracking_uri()).scheme
```
Comment (harupy) [inline examples/tensorflow/train.py:189]:
```suggestion
```

File: examples/tensorflow/train.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+            train_loss = tf.reduce_mean(batch_losses_train)
+            test_loss = tf.reduce_mean(batch_losses_test)
+            train_losses.append(train_loss)
+            test_losses.append(test_loss)
+            if epoch % 10 == 0:
+                mlflow.log_metric(key="train_losses", value=train_loss, step=epoch)
+                mlflow.log_metric(key="test_losses", value=test_loss, step=epoch)
+                print(f"Mean squared error for step {epoch}: {train_loss.numpy():0.3f}")
+
+        # Log the parameters
```
Comment (harupy) [inline examples/tensorflow/train.py:174]:
```suggestion
    with mlflow.start_run():
        # Log the parameters
```

File: examples/tensorflow/train.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    x_test, y_test = test_data[:, 1:], test_data[:, 0]
+    # One hot encode origin column
+    x_train_ohe, x_test_ohe = onehot_origin(x_train), onehot_origin(x_test)
+    # Data normalization
+    norm_x = Normalize(x_train_ohe)
+    norm_y = Normalize(y_train)
+    x_train_norm, y_train_norm = norm_x.norm(x_train_ohe), norm_y.norm(y_train)
+    x_test_norm, y_test_norm = norm_x.norm(x_test_ohe), norm_y.norm(y_test)
+
+    with mlflow.start_run():
```
Comment (harupy) [inline examples/tensorflow/train.py:125]:
```suggestion
```

File: examples/tensorflow/train.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            test_losses.append(test_loss)
+            if epoch % 10 == 0:
+                mlflow.log_metric(key="train_losses", value=train_loss, step=epoch)
+                mlflow.log_metric(key="test_losses", value=test_loss, step=epoch)
+                print(f"Mean squared error for step {epoch}: {train_loss.numpy():0.3f}")
+
+        # Log the parameters
+        mlflow.log_param("epochs", epochs)
+        mlflow.log_param("learning_rate", learning_rate)
+        mlflow.log_param("batch_size", batch_size)
```
Comment (harupy) [inline examples/tensorflow/train.py:177]:
```suggestion
        mlflow.log_params({
            "epochs": epochs,
            "learning_rate": learning_rate,
            "batch_size", batch_size,
        })
```

File: examples/tensorflow/train.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        # Log the final metrics
+        mlflow.log_metric("final_train_loss", train_loss)
+        mlflow.log_metric("final_test_loss", test_loss)
```
Comment (harupy) [inline examples/tensorflow/train.py:180]:
```suggestion
        mlflow.log_metrics(...)
```

File: examples/tensorflow/train.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    column_names = [
+        "MPG",
+        "Cylinders",
+        "Displacement",
+        "Horsepower",
+        "Weight",
+        "Acceleration",
+        "Model Year",
+        "Origin",
+    ]
```
Comment (harupy) [inline examples/tensorflow/train.py:104]: Can we use [the California housing dataset](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_california_housing.html )
instead?

File: examples/tensorflow/train.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    return tf.reduce_mean(tf.square(y_pred - y))
+
+
+# Enable auto-logging to MLflow to capture TensorBoard metrics.
+mlflow.tensorflow.autolog()
+
+if __name__ == "__main__":
+    # Set a random seed for reproducible results
+    tf.random.set_seed(42)
+
```
Comment (harupy) [inline examples/tensorflow/train.py:92]:
```suggestion
# Enable auto-logging to MLflow to capture TensorBoard metrics.

if __name__ == "__main__":
    # Set a random seed for reproducible results
    tf.random.set_seed(42)

    mlflow.tensorflow.autolog()
```

**Comment by harupy** [comment]
@dheerajnbha thanks for the PR, left some comments!

File: examples/tensorflow/train.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+import logging
+from urllib.parse import urlparse
+
+import mlflow
+import pandas as pd
+
+import tensorflow as tf
+
+logging.basicConfig(level=logging.WARN)
+logger = logging.getLogger(__name__)
```
Comment (harupy) [inline examples/tensorflow/train.py:11]: Do we use logger in this example?

File: examples/tensorflow/train.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+if __name__ == "__main__":
+    # Set a random seed for reproducible results
+    tf.random.set_seed(42)
+
+    # Enable auto-logging to MLflow to capture TensorBoard metrics.
+    mlflow.tensorflow.autolog()
+
+    # Dataset url
+    url = "https://raw.githubusercontent.com/dheerajnbhat/datasets/main/ml/californiaHousing/californiaHousing.csv"
+    dataset = pd.read_csv(url, na_values="?")
```
Comment (harupy) [inline examples/tensorflow/train.py:83]: Can we use:

```
from sklearn.datasets import fetch_california_housing
```

instead?

File: examples/tensorflow/train.py (hunk 2 above)
Comment (harupy) [reply examples/tensorflow/train.py:1]: Thanks!

File: examples/tensorflow/train.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
+# tensorflow 2.x core api
+import logging
+from urllib.parse import urlparse
```
Comment (harupy) [inline examples/tensorflow/train.py:3]: Let's remove this. It's not used.

File: examples/tensorflow/conda.yaml [hunk 14]
```diff
@@ -0,0 +1,11 @@
+name: tutorial
+channels:
+  - conda-forge
+  - defaults
```
Comment (harupy) [inline examples/tensorflow/conda.yaml:4]:
```suggestion
```

File: examples/tensorflow/conda.yaml (hunk 1 above)
Comment (harupy) [inline examples/tensorflow/conda.yaml:11]:
```suggestion
```

mlflow depends on pandas, this can be removed

**Review by harupy — APPROVED** [review]
Left a few more comments, otherwise LGTM!

File: examples/tensorflow/train.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+    """Calculating Mean Square Error Loss function"""
+    return tf.reduce_mean(tf.square(y_pred - y))
+
+
+if __name__ == "__main__":
+    # Set a random seed for reproducible results
+    tf.random.set_seed(42)
+
+    # Enable auto-logging to MLflow to capture TensorBoard metrics.
+    mlflow.tensorflow.autolog()
```
Comment (harupy) [inline examples/tensorflow/train.py:79]:
```suggestion
```

It looks like autologging does nothing in this example. Can we remove it?

File: examples/tensorflow/train.py (hunk 13 above)
Comment (dheerajnbhat) [reply examples/tensorflow/train.py:3]: Yeah, urllib was used earlier when I had used url tracking type, now its not needed

File: examples/tensorflow/train.py (hunk 11 above)
Comment (dheerajnbhat) [reply examples/tensorflow/train.py:11]: Not using now, will remove it

File: examples/tensorflow/train.py [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+class Normalize(tf.Module):
+    """Data Normalization class"""
+
+    def __init__(self, x):
+        # Initialize the mean and standard deviation for normalization
+        self.mean = tf.math.reduce_mean(x, axis=0)
+        self.std = tf.math.reduce_std(x, axis=0)
+
+    def norm(self, x):
+        # Normalize the input
```
Comment (harupy) [inline examples/tensorflow/train.py:23]:
```suggestion
```

Let's remove this. It's obvious we're normalizing the input here.

File: examples/tensorflow/train.py [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+        return (x - self.mean) / self.std
+
+    def unnorm(self, x):
+        # Unnormalize the input
```
Comment (harupy) [inline examples/tensorflow/train.py:27]:
```suggestion
```

Same here

File: examples/tensorflow/train.py (hunk 12 above)
Comment (dheerajnbhat) [reply examples/tensorflow/train.py:81]: Since it is tensorflow example, I did not want to add any other dependency except tensorflow and mlflow. The dataset that is being downloaded has been taken from the sklearn.datasets.fetch_california_housing() only.

File: examples/tensorflow/python_env.yaml [hunk 18]
```diff
@@ -0,0 +1,7 @@
+python: "3.8"
+build_dependencies:
+  - pip
+dependencies:
+  - mlflow>=2.0
+  - tensorflow>=2.8
+  - pandas
```
Comment (harupy) [inline examples/tensorflow/python_env.yaml:7]:
```suggestion
```

File: examples/tensorflow/train.py (hunk 12 above)
Comment (harupy) [reply examples/tensorflow/train.py:72]: Got it, but mlflow depends on sklearn.

File: examples/tensorflow/train.py [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+            }
+        )
+        print(f"\nFinal train loss: {train_loss:0.3f}")
+        print(f"Final test loss: {test_loss:0.3f}")
+
+        # Export the tensorflow model
+        lin_reg_export = ExportModule(model=lin_reg, norm_x=norm_x, norm_y=norm_y)
+
+        # Model registry does not work with file store
+        # I think we can just log the model.
```
Comment (harupy) [inline examples/tensorflow/train.py:161]:
```suggestion
```

File: examples/tensorflow/train.py (hunk 12 above)
Comment (dheerajnbhat) [reply examples/tensorflow/train.py]: Yeah, true ;). I have changed from url to sklearn.datasets.

File: examples/tensorflow/train.py [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+    """Calculating Mean Square Error Loss function"""
+    return tf.reduce_mean(tf.square(y_pred - y))
+
+
+if __name__ == "__main__":
+    # Set a random seed for reproducible results
+    tf.random.set_seed(42)
+
+    # Load dataset
+    dataset = pd.DataFrame(data=fetch_california_housing(as_frame=True)["frame"])
```
Comment (harupy) [inline examples/tensorflow/train.py:72]:
```suggestion
    dataset = fetch_california_housing(as_frame=True)["frame"]
```

Do we need `pd.DataFrame`?

```python
>>> type(fetch_california_housing(as_frame=True)["frame"])
<class 'pandas.core.frame.DataFrame'>
```

**Comment by harupy** [comment]
@dheerajnbhat Merged, thanks for the PR!

**Comment by dheerajnbhat** [comment]
> @dheerajnbhat Merged, thanks for the PR!

Glad to contribute to mlflow :)
